### PR TITLE
GPTJ Tokenizer Supported Thai Token

### LINKED ISSUE #76: [Super AI] [Good First Issue] Convert Pretrained GPTJ to Support Thai Token
Description:
There is a [working example](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md) of continue pertaining LLM to support other languages (In their case Chinese). We will try to do the same starting by extending the vocabulary of GPTJ to the Thai Language.

Instructions:
1. Since GPTJ uses BPE tokenizer, we will stick with [GPT-base-thai](https://huggingface.co/flax-community/gpt2-base-thai) Tokenizer for fast iteration
2. Develop a conversion script to merge GPT-base-thai Tokenizer Vocab to GPTJ tokenizer
3. Test the merged GPTJ tokenizer. Expected to still have similar tokenized output when running with English sentence.

Deliverables:
- Convert/Inference Code Script
- PR
- Unit Test

### PR DESCRIPTION
**Pull request by Pattptr**
merge [GPTJ Tokenizer](https://huggingface.co/EleutherAI/gpt-j-6b/tree/main) with [gpt2-base-thai](https://huggingface.co/flax-community/gpt2-base-thai/tree/main) Tokenizer by extent vocabulary following [working example](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md) 

- To test merged method
```pip install -e ./src/model```
1.  need to have [GPTJ Tokenizer](https://huggingface.co/EleutherAI/gpt-j-6b/tree/main) and [gpt2-base-thai](https://huggingface.co/flax-community/gpt2-base-thai/tree/main) 
2.  run tokenizer.py
``` python3 /root/openthaigpt-pretraining/src/model/openthaigpt_pretraining_model/models/GPTJ_TH_tokenizer/tokenizer.py```

## Related Issues
Close #76

### CODE REVIEW

File: src/model/openthaigpt_pretraining_model/models/GPTJ_TH_tokenizer/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        return self.tokenizer.tokenize(x)
+
+    def encode(self, x: str) -> List[Any]:
+        return self.tokenizer.encode(x)
+
+    def decode(self, x: Union[int, List[int]]) -> str:
+        return self.tokenizer.decode(x)
+
+
+text = "รายละเอียดและหลักเกณฑ์การคัดเลือก AI Startup Incubation by AIEAT"
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/GPTJ_TH_tokenizer/tokenizer.py:34]: You should put the code below this to src/model/scripts. Then you can create you own folder. After you move it, please add README.md and describe what you do.

Comment (Pattptr) [reply src/model/openthaigpt_pretraining_model/models/GPTJ_TH_tokenizer/tokenizer.py]: I' ve commented lines below and move them to the test case in tests/model and I've move the token merging script to src/model/scripts and also add README.md file in that folder.

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    gptj_tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")
+    gpt2_tokenizer = AutoTokenizer.from_pretrained("flax-community/gpt2-base-thai")
+
+    # load vcab and merge rule
+    hf_hub_download(
+        repo_id="EleutherAI/gpt-j-6b",
+        filename="merges.txt",
+        local_dir=f"{FILE_DIR}/GPTJ_merge_rule/",
+    )
+    hf_hub_download(
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py:18]: Can we make repo_id customizable `def merge(thai_tokenizer_repo: str)`

```
hf_hub_download(
        repo_id=thai_tokenizer_repo,
        filename="merges.txt",
        local_dir=f"{FILE_DIR}/GPT2_merge_rule/",
    )
```

File: src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    for word in gpt2_vocab.keys():
+        if word not in new_vocab.keys():
+            new_vocab[word] = idx
+            idx += 1
+
+    # convert dictionary to json
+    new_vocab_json = json.dumps(new_vocab, ensure_ascii=False)
+
+    # write json to a file
+    with open("merge_vocab.json", "w", encoding="utf-8") as outfile:
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py:46]: Can you create folder called `/temp` on the root directory and save merge_vocab.json there

Please do `.gitignore` for `/temp`

File: src/model/scripts/gptj_thai_tokenizer/merge_tokenizers.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from openthaigpt_pretraining_model.GPTJ_TH_tokenizer.merge import merge
+from openthaigpt_pretraining_model.GPTJ_TH_tokenizer.constants import (
+    OUTPUT_HF_DIR,
+)
+
+tokenizer = merge()
```
Comment (new5558) [inline src/model/scripts/gptj_thai_tokenizer/merge_tokenizers.py:6]: Can you make thai_tokenizer available as an argument when running this file? If the user did not selected thai_tokenzier the default should be `flax-community/gpt2-base-thai`

Example:
```python
import argparse

parser = argparse.ArgumentParser(description='Process some integers.')

parser.add_argument('integers', metavar='N', type=int, nargs='+',
                    help='an integer for the accumulator')

parser.add_argument('--sum', dest='accumulate', action='store_const',
                    const=sum, default=max,
                    help='sum the integers (default: find the max)')

args = parser.parse_args()
print(args.accumulate(args.integers))
```

File: src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py (hunk 2 above)
Comment (Pattptr) [reply src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py:20]: Done

File: src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py (hunk 3 above)
Comment (Pattptr) [reply src/model/openthaigpt_pretraining_model/GPTJ_TH_tokenizer/merge.py]: Done

File: src/model/scripts/gptj_thai_tokenizer/merge_tokenizers.py (hunk 4 above)
Comment (Pattptr) [reply src/model/scripts/gptj_thai_tokenizer/merge_tokenizers.py]: Done

**Review by new5558 — APPROVED** [review]

**Pull request merged**